sdk/src/types/block/protocol.rs

```diff
@@ -224,6 +224,11 @@ impl Default for WorkScoreStructure {
     }
 }
 
+/// A trait to facilitate the computation of the work score of a block, which is central to mana cost calculation.
+pub trait WorkScore {
+    fn workscore(&self, workscore_structure: WorkScoreStructure) -> u32;
```

Comment:
maybe something like this instead
```rust
pub struct WorkScoreBuilder {
    total: u32,
    structure: WorkScoreStructure,
}

impl WorkScoreBuilder {
    pub fn new(structure: WorkScoreStructure) -> Self {
        Self {
            total: 0,
            structure,
        }
    }
}

pub trait WorkScore {
    fn build_workscore(&self, builder: WorkScoreBuilder) -> WorkScoreBuilder;

    fn workscore(&self, workscore_structure: WorkScoreStructure) -> u32 {
        self.build_workscore(WorkScoreBuilder::new(workscore_structure)).total
    }
}

Reply:
I think I understand where you're getting at. "...more useful fns" are basically `with_` builder methods?

Reply:
tbh ... I don't think there's a need for a builder here, but maybe @thibault-martinez sees it differently. The only thing that would probably make sense to add to the trait is a pre-implemented `mana_cost` fn. Then the `StorageScore` trait (#1398) and the `WorkScore` trait would be consistent as can be IMO.